File: app/src/main/java/com/example/a_place/PanelWeb.java
package com.example.a_place;

import android.app.Activity;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.opengl.GLES20;
import android.opengl.GLUtils;
import android.util.Log;
import android.view.View;
import android.webkit.WebView;
import android.webkit.WebViewClient;

public class PanelWeb {
    public int textureId = -1;
    public float xOffset = 0f;
    public float yOffset = 0f;

    private final Context context;
    private final String url;
    private final int altura;
    private final int anchura;

    private WebView webView;
    private boolean isReady = false;
    private Bitmap pendingBitmap = null;
    private final Object bitmapLock = new Object();
    private boolean needsUpdate = false;

    public PanelWeb(Context context, String url, int width, int height) {
        this.context = context;
        this.url = url;
        this.altura = width;
        this.anchura = height;

        if (context instanceof Activity) {
            ((Activity) context).runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    webView = new WebView(context);
                    webView.setLayerType(View.LAYER_TYPE_SOFTWARE, null);

                    webView.getSettings().setJavaScriptEnabled(true);
                    //webView.getSettings().setDomStorageEnabled(true); no creo que se ocupa la verdad
                    webView.getSettings().setUseWideViewPort(true);
                    webView.getSettings().setLoadWithOverviewMode(true);

                    webView.setWebViewClient(new WebViewClient() {
                        @Override
                        public void onPageFinished(WebView view, String urlString) {
                            isReady = true;
                            Log.d("PanelWeb", "CARGO !!! " + urlString);
                            triggerSnapshot();
                        }
                    });

                    webView.layout(0, 0, width, height); // aver aca segun yo se pone en el ui talvez me quivoque
                    int specancho = View.MeasureSpec.makeMeasureSpec(width, View.MeasureSpec.EXACTLY);
                    int specalto = View.MeasureSpec.makeMeasureSpec(height, View.MeasureSpec.EXACTLY);
                    webView.measure(specancho, specalto);

                    webView.loadUrl(url);
                }
            });
        }
    }

    public void iniciarTextura() {
        int[] tex = new int[1];
        GLES20.glGenTextures(1, tex, 0);
        textureId = tex[0];

        GLES20.glBindTexture(GLES20.GL_TEXTURE_2D, textureId);
        GLES20.glTexParameteri(GLES20.GL_TEXTURE_2D, GLES20.GL_TEXTURE_MIN_FILTER, GLES20.GL_LINEAR);
        GLES20.glTexParameteri(GLES20.GL_TEXTURE_2D, GLES20.GL_TEXTURE_MAG_FILTER, GLES20.GL_LINEAR);
        GLES20.glTexParameteri(GLES20.GL_TEXTURE_2D, GLES20.GL_TEXTURE_WRAP_S, GLES20.GL_CLAMP_TO_EDGE);
        GLES20.glTexParameteri(GLES20.GL_TEXTURE_2D, GLES20.GL_TEXTURE_WRAP_T, GLES20.GL_CLAMP_TO_EDGE);

        Log.d("PanelWeb", "GL_TEXTURE_2D inicio con ID: " + textureId);
    }

    private void triggerSnapshot() {
        if (!isReady || webView == null) return;

        try {
            Bitmap bmp = Bitmap.createBitmap(anchura, altura, Bitmap.Config.ARGB_8888);
            Canvas canvas = new Canvas(bmp);
            canvas.drawColor(Color.WHITE);
            webView.draw(canvas);

            synchronized (bitmapLock) {
                if (pendingBitmap != null) {
                    pendingBitmap.recycle();
                }
                pendingBitmap = bmp;
                needsUpdate = true;
            }
        } catch (Exception e) {
            Log.e("PanelWeb", "error al convertir el webview a bitmap: " + e.getMessage(), e);
        }
    }

    public void update() {
        if (isReady && !needsUpdate) {
            if (context instanceof Activity) {
                ((Activity) context).runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        triggerSnapshot();
                    }
                });
            }
        }
    }

    public void updateTexture() {
        if (textureId == -1) return;

        Bitmap bmp = null;
        synchronized (bitmapLock) {
            if (needsUpdate) {
                bmp = pendingBitmap;
                pendingBitmap = null;
                needsUpdate = false;
            }
        }

        if (bmp != null) {
            GLES20.glBindTexture(GLES20.GL_TEXTURE_2D, textureId);
            GLUtils.texImage2D(GLES20.GL_TEXTURE_2D, 0, bmp, 0);
            bmp.recycle();
        }
    }
}

File: app/src/main/java/com/example/a_place/PanelWeb.kt
package com.example.a_place

import android.app.Activity
import android.content.Context
import android.graphics.SurfaceTexture
import android.opengl.GLES11Ext
import android.opengl.GLES20
import android.view.Surface
import android.view.View
import android.webkit.WebView
import android.webkit.WebViewClient

class PanelWeb(val context: Context, val url: String, val width: Int, val height: Int) {
    var textureId: Int = -1
    var surfaceTexture: SurfaceTexture? = null
    var surface: Surface? = null
    var webView: WebView? = null
    var xOffset: Float = 0f
    var yOffset: Float = 0f

    init {
        (context as Activity).runOnUiThread {
            webView = WebView(context).apply {
                settings.javaScriptEnabled = true
                settings.domStorageEnabled = true
                setLayerType(View.LAYER_TYPE_HARDWARE, null)
                layout(0, 0, width, height)
                measure(View.MeasureSpec.makeMeasureSpec(width, View.MeasureSpec.EXACTLY),
                    View.MeasureSpec.makeMeasureSpec(height, View.MeasureSpec.EXACTLY))
                webViewClient = WebViewClient()

                // FIX: Use the class property explicitly
                loadUrl(this@PanelWeb.url)
            }
        }
    }

    fun iniciarTextura() {
        val tex = IntArray(1)
        GLES20.glGenTextures(1, tex, 0)
        textureId = tex[0]

        // GL_TEXTURE_EXTERNAL_OES (0x8D65)
        val type = GLES11Ext.GL_TEXTURE_EXTERNAL_OES
        GLES20.glBindTexture(type, textureId)

        // parametros para prevenir pantalla negra
        GLES20.glTexParameteri(type, GLES20.GL_TEXTURE_MIN_FILTER, GLES20.GL_LINEAR)
        GLES20.glTexParameteri(type, GLES20.GL_TEXTURE_MAG_FILTER, GLES20.GL_LINEAR)
        GLES20.glTexParameteri(type, GLES20.GL_TEXTURE_WRAP_S, GLES20.GL_CLAMP_TO_EDGE)
        GLES20.glTexParameteri(type, GLES20.GL_TEXTURE_WRAP_T, GLES20.GL_CLAMP_TO_EDGE)

        surfaceTexture = SurfaceTexture(textureId)
        surfaceTexture?.setDefaultBufferSize(width, height)
        surface = Surface(surfaceTexture)
    }

    // Inside PanelWeb.kt
    var needsUpdate = false // Add this flag

    fun update() {
        (context as? Activity)?.runOnUiThread {
            try {
                val canvas = surface?.lockCanvas(null)
                if (canvas != null) {
                    webView?.draw(canvas)
                    surface?.unlockCanvasAndPost(canvas)
                    needsUpdate = true // Signal that new data is ready
                }
            } catch (e: Exception) { e.printStackTrace() }
        }
    }

    // Add this function to be called in onDrawFrame
    fun updateTexture() {
        if (needsUpdate) {
            try {
                surfaceTexture?.updateTexImage()

                needsUpdate = false // Reset flag
            } catch (e: Exception) {}
        }
    }
}